titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/controller/NodeGcController.java

```diff
@@ -0,0 +1,180 @@
+
+    private final KubeControllerConfiguration configuration;
+    private final AgentManagementService agentManagementService;
+    private final KubeApiFacade kubeApiFacade;
+    private final LocalScheduler scheduler;
+
+    private final Clock clock;
+    private final Gauge nodesToGc;
+
+    @Inject
+    public NodeGcController(KubeControllerConfiguration configuration,
+                            TitusRuntime titusRuntime,
+                            AgentManagementService agentManagementService,
+                            KubeApiFacade kubeApiFacade,
+                            LocalScheduler scheduler) {
+        this.configuration = configuration;
+        this.agentManagementService = agentManagementService;
+        this.kubeApiFacade = kubeApiFacade;
+        this.scheduler = scheduler;
+
+        this.clock = titusRuntime.getClock();
+        this.nodesToGc = titusRuntime.getRegistry().gauge(METRIC_ROOT + "nodesToGc");
+    }
+
+    @Activator
+    public void enterActiveMode() {
+        ScheduleDescriptor reconcileSchedulerDescriptor = ScheduleDescriptor.newBuilder()
+                .withName("nodeGcController")
+                .withDescription("GC nodes that no longer exist")
+                .withInitialDelay(Duration.ofMillis(configuration.getNodeGcControllerInitialDelayMs()))
+                .withInterval(Duration.ofMillis(configuration.getNodeGcControllerIntervalMs()))
+                .withTimeout(Duration.ofMillis(configuration.getNodeGcControllerTimeoutMs()))
+                .build();
+        scheduler.schedule(reconcileSchedulerDescriptor, e -> gcNodes(), ExecutorsExt.namedSingleThreadExecutor("node-gc-controller"));
+    }
+
+    void gcNodes() {
+        if (!kubeApiFacade.getNodeInformer().hasSynced()) {
+            logger.info("Node informer has not synced");
+            return;
+        }
+
+        if (!configuration.isNodeGcControllerEnabled()) {
+            logger.info("Node GC is disabled");
+            return;
+        }
+
+        List<V1Node> nodes = kubeApiFacade.getNodeInformer().getIndexer().list();
+        Set<String> nodeNamesToGc = nodes.stream()
+                .filter(this::isNodeEligibleForGc)
+                .map(n -> KubeUtil.getMetadataName(n.getMetadata()))
+                .filter(StringExt::isNotEmpty)
+                .collect(Collectors.toSet());
+
+        logger.info("Attempting to GC {} nodes: {}", nodeNamesToGc.size(), nodeNamesToGc);
+        this.nodesToGc.set(nodeNamesToGc.size());
+        for (String nodeName : nodeNamesToGc) {
+            gcNode(nodeName);
+        }
+        logger.info("Finished node GC");
+    }
+
+    @VisibleForTesting
+    boolean isNodeEligibleForGc(V1Node node) {
+        String nodeName = KubeUtil.getMetadataName(node.getMetadata());
+        Optional<V1NodeCondition> readyNodeConditionOpt = KubeUtil.findNodeCondition(node, READY);
+        if (StringExt.isNotEmpty(nodeName) && readyNodeConditionOpt.isPresent()) {
```

Comment:
I am wondering how likely is the scenario where nodeName is missing/empty and if that is indicative of a different issue.

Reply:
This should not happen but the models/API don't provide any guarantees so there is a lot of defensive checks.